### ISSUE TITLE
代码权值没有保存好像

### ISSUE THREAD

**Issue by Issac304** [Sampson-Lee/DSCT#5]
运行测试时候有问题

**Comment by Issac304** [comment]
运行test.py时候权重文件加载不出来
![image](https://github.com/user-attachments/assets/ee60d359-2859-4321-9daa-fc9ad079e2a4)

**Comment by Sampson-Lee** [comment]
Hi, thank you for your feedback.

I think you were using the pretrained weights of previous code. 

I tried the latest code for training and testing and everything ran smoothly.

**Comment by Issac304** [comment]
![image](https://github.com/user-attachments/assets/ab93e07c-8c40-4650-888e-be0ac01523e6)
我记得我是重新拉后训练的，应该是更新的test有问题把？之前emotic都不能了

**Comment by Issac304** [comment]
键是fetr他要求detr

**Comment by Sampson-Lee** [comment]
Thank you for your feedback. Would you like to try the following two options:

Option 1 (Recommended): train the task again with 2 epochs and then test. I still think you were using the previous code for training since the latest code does not contain fetr but the previous code does.

```
# #####emotic training scripts #####
YOUR_DATA_PATH=/home/lxp/data/emotic
CUDA_VISIBLE_DEVICES=0 python -m torch.distributed.launch \
    --nproc_per_node=1 \
    --master_port=29507 \
    --use_env main.py \
    --dataset_file=emotic \
    --binary_flag=1 \
    --detr=deformable_detr_dsct \
    --model=deformable_transformer_dsct \
    --batch_size=4 \
    --cls_loss_coef=5 \
    --data_path=$YOUR_DATA_PATH \
    --output_dir=$YOUR_DATA_PATH/checkpoints \
    --epochs=2 \ // Adjust it to 2
    --lr_drop=1 \ // Adjust it to 1
    --num_queries=4;
```

Option 2: rename inconsistent keys. Here is the code for your reference.

```
import torch

# Load pretrained weights
pretrained_weights = torch.load("pretrained_model.pth", map_location="cpu")

# Renaming keys in pretrained weights
renamed_weights = {}
for key, value in pretrained_weights.items():
    new_key = key.replace("fetr", "dsct")  # Replace fetr with dsct or other mismatch
    renamed_weights[new_key] = value

# Load renamed weights into model
model.load_state_dict(renamed_weights, strict=False)
print("Loaded weights after renaming keys.")

```

**Comment by Issac304** [comment]
![image](https://github.com/user-attachments/assets/789b23bc-190d-488f-9ead-7e60c1dee7ec)
按照您的方案1  重新训练了  确实emotic好了   但是caer还有这个问题  不知怎么回事

**Comment by Issac304** [comment]
我是把本地代码删除，重新拉，然后更新到服务器，然后重新训，emotic完美解决了，caer还有瑕疵

**Comment by Sampson-Lee** [comment]
Hi,

Thank you for the feedback!

Did you set `num_queries` to 4 during training on the CAER dataset and to 9 during testing?

Everything works correctly on my end.

**Comment by Issac304** [comment]
确实是您说的那个问题，解决了，但是还有另外两个问题

1.不能恢复断点继续训练
![image](https://github.com/user-attachments/assets/cb16cf75-fedf-4976-b18e-d11453883600)
最下面那个是恢复的，感觉好像权值没有加载进来，和之前差一大截

2.caer的val用的是test的数据？那训练时候就见过test吗？emotic 的val就是val好像，不解
![image](https://github.com/user-attachments/assets/5fcb0a7c-3926-42ce-b4b1-44b1627c9d86)

**Comment by Sampson-Lee** [comment]
Q1. 不能恢复断点继续训练

A1. Yes, you are correct. There was a bug in main.py where the pretrained weights were overriding the resumed weights. I have fixed this issue in the latest version of the code.

Q2. caer的val用的是test的数据？那训练时候就见过test吗？emotic 的val就是val好像，不解

A2. I mistakenly used the test set to fill the validation set, but this does not impact the fairness of training since the test set was only used for evaluation, not for training. To resolve this confusion, I have uploaded the proper validation annotations and updated caer.py.

**Comment by Issac304** [comment]
确实是覆盖了的问题，解决了。但是测试又有问题了，我是重新训2个epoch再测试的（您之前推荐方法），有bug
![image](https://github.com/user-attachments/assets/86b85a25-bf90-49aa-ad72-193c6c0f8d60)

**Comment by Issac304** [comment]
我是重新拉了代码

**Comment by Sampson-Lee** [comment]
My apologies for the oversight. I updated the code in test.py but forgot to notify you. The change was made to ensure consistency between the training and testing logic.

In the previous version, the argument resume was used for loading weights. In the updated version, the argument pretrained_weights is now used for this purpose.

As a result, the testing script should now be as follows:

```
##### Emotic Testing Script #####
YOUR_DATA_PATH=/home/lxp/data/emotic
YOUR_MODEL_PATH=/home/lxp/data/emotic/checkpoints
CUDA_VISIBLE_DEVICES=0 python test.py \
        --dataset_file=emotic \
        --detr=deformable_detr_dsct \
        --model=deformable_transformer_dsct \
        --num_queries=4 \
        --binary_flag=1 \
        --data_path=$YOUR_DATA_PATH/images \
        --json_path=./datasets/annotations/emotic_test_bi.json \
        --pretrained_weights=$YOUR_MODEL_PATH/checkpoint.pth;
```

Let me know if you encounter any issues!

**Comment by Issac304** [comment]
完美解决了，太感谢了！

**Comment by Issac304** [comment]
完美解决了，丝滑运行起来了，谢谢同学了，同学太nice了




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Sampson-Lee/DSCT"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年12月18日(星期三) 上午10:41
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [Sampson-Lee/DSCT] 代码权值没有保存好像 (Issue #5)





 
My apologies for the oversight. I updated the code in test.py but forgot to notify you. The change was made to ensure consistency between the training and testing logic.
 
In the previous version, the argument resume was used for loading weights. In the updated version, the argument pretrained_weights is now used for this purpose.
 
As a result, the testing script should now be as follows:
 ##### Emotic Testing Script ##### YOUR_DATA_PATH=/home/lxp/data/emotic YOUR_MODEL_PATH=/home/lxp/data/emotic/checkpoints CUDA_VISIBLE_DEVICES=0 python test.py \         --dataset_file=emotic \         --detr=deformable_detr_dsct \         --model=deformable_transformer_dsct \         --num_queries=4 \         --binary_flag=1 \         --data_path=$YOUR_DATA_PATH/images \         --json_path=./datasets/annotations/emotic_test_bi.json \         --pretrained_weights=$YOUR_MODEL_PATH/checkpoint.pth;  
Let me know if you encounter any issues!
